src/core/ad.jl

```diff
@@ -102,43 +106,54 @@ function gradient_logp_forward(
     chunk = ForwardDiff.Chunk(min(length(θ), chunk_size))
     config = ForwardDiff.GradientConfig(f, θ, chunk)
     ∂l∂θ = ForwardDiff.gradient!(similar(θ), f, θ, config)
-    l = vi.logp
-    vi.logp = logp_old
+    l = getlogp(vi)
+    setlogp!(vi, logp_old)
 
     return l, ∂l∂θ
 end
 
 """
 gradient_logp_reverse(
+    backend::ADBackend,
     θ::AbstractVector{<:Real},
     vi::VarInfo,
     model::Model,
-    sampler::AbstractSampler=SampleFromPrior(),
+    sampler::AbstractSampler = SampleFromPrior(),
 )
 
 Computes the value of the log joint of `θ` and its gradient for the model
-specified by `(vi, sampler, model)` using reverse-mode AD from Tracker.jl.
+specified by `(vi, sampler, model)` using reverse-mode AD from the specified `backend`, e.g. `TrackerAD()` which uses `Tracker.jl` or `ZygoteAD()` which uses `Zygote.jl`.
 """
 function gradient_logp_reverse(
+    backend::TrackerAD,
     θ::AbstractVector{<:Real},
     vi::VarInfo,
     model::Model,
-    sampler::AbstractSampler=SampleFromPrior(),
+    sampler::AbstractSampler = SampleFromPrior(),
 )
-    T = typeof(vi.logp)
+    T = typeof(getlogp(vi))
 
     # Specify objective function.
     function f(θ)
         new_vi = VarInfo(vi, sampler, θ)
-        return runmodel!(model, new_vi, sampler).logp
+        return getlogp(runmodel!(model, new_vi, sampler))
     end
 
     # Compute forward and reverse passes.
     l_tracked, ȳ = Tracker.forward(f, θ)
-    l::T, ∂l∂θ::typeof(θ) = Tracker.data(l_tracked), Tracker.data(ȳ(1)[1])
     # Remove tracking info from variables in model (because mutable state).
+    l::T, ∂l∂θ::typeof(θ) = Tracker.data(l_tracked), Tracker.data(ȳ(1)[1])
+
     return l, ∂l∂θ
 end
+function gradient_logp_reverse(
```

Comment:
Is this function (without explicit backend) actually needed? If only `gradient_logp` is called in the implemented algorithms, then the method with explicit backend is called anyways all the time.

Reply:
It's only called in tests I think.

Reply:
It's convenient when poking around AD bugs so let's keep it for now.